File: cses/02_sorting_searching/2162_josephus-problem-i/solution.py
import sys
from collections import deque


def main():
    # Đọc toàn bộ input một lần rồi tách thành các token để tăng tốc I/O.
    data = sys.stdin.buffer.read().split()
    n = int(data[0])

    # Đưa tất cả n đứa trẻ vào deque theo thứ tự 1..n để mô phỏng vòng tròn.
    q = deque(range(1, n + 1))
    res = []
    while q:
        # Bỏ qua (skip): chuyển phần tử đầu xuống cuối hàng đợi.
        q.append(q.popleft())
        # Loại (remove): phần tử đầu tiếp theo bị loại, ghi vào kết quả.
        res.append(q.popleft())

    # Gộp toàn bộ kết quả thành một chuỗi rồi ghi ra một lần để tăng tốc I/O.
    sys.stdout.write(' '.join(map(str, res)) + '\n')


main()

File: cses/02_sorting_searching/2162_josephus-problem-i/solution.cpp
#include <deque>
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n;
    std::cin >> n;

    // Đưa tất cả n đứa trẻ vào deque theo thứ tự 1..n để mô phỏng vòng tròn.
    std::deque<int> children;
    for (int child = 1; child <= n; ++child) {
        children.push_back(child);
    }

    // Lặp cho tới khi deque rỗng: bỏ qua đứa đầu (đưa xuống cuối) rồi loại đứa tiếp theo.
    std::vector<int> order;
    order.reserve(n);
    while (!children.empty()) {
        // Bỏ qua (skip): chuyển phần tử đầu xuống cuối hàng, nó vẫn còn trong vòng.
        children.push_back(children.front());
        children.pop_front();
        // Loại (remove): phần tử đầu tiếp theo bị loại, ghi vào thứ tự kết quả.
        order.push_back(children.front());
        children.pop_front();
    }

    // Xuất thứ tự bị loại, các số cách nhau bởi dấu cách.
    for (int i = 0; i < n; ++i) {
        if (i > 0) {
            std::cout << ' ';
        }
        std::cout << order[i];
    }
    std::cout << '\n';

    return 0;
}
